lib/src/css_mirroring/util_classes.dart

```diff
@@ -0,0 +1,69 @@
+part of scissors.src.css_mirroring.bidi_css_generator;
+
+/// Used to decide which part of CSS to retain in a transaction.
+enum _RetentionMode {
+  /// Keep parts of CSS which are direction-independent eg: color and width.
+  keepBidiNeutral,
+
+  /// Keep direction dependent parts of original CSS eg: margin.
+  keepOriginalBidiSpecific,
+
+  /// to keep direction dependent parts of flipped CSS.
+  keepFlippedBidiSpecific
+}
+
+
+/// Stores the start and end locations of the pending removals.
+class _PendingRemovals {
+  final String source;
+  final TextEditTransaction _transaction;
+
+  // List to contain start and end location of pending removals.
+  final _removalStartEndLocations = <_Pair>[];
+
+  _PendingRemovals(TextEditTransaction trans)
+      : _transaction = trans,
+        source = trans.file.getText(0);
+
+  void addRemoval(int start, int end) {
+    _removalStartEndLocations.add(new _Pair(start, end));
+  }
+
+  void commitRemovals() {
+    _removalStartEndLocations.forEach((_Pair p) {
+      _transaction.edit(p.first, p.second, '');
+    });
+    _removalStartEndLocations.clear();
+  }
+}
+
+class _Pair<T> {
+  final T first;
+  final T second;
+
+  const _Pair(this.first, this.second);
+}
+
+class TopLevelsPair {
+  final List<TreeNode> _originalTopLevels;
+  final List<TreeNode> _flippedTopLevels;
+
+  TopLevelsPair(this._originalTopLevels, this._flippedTopLevels) {
+    assert(_originalTopLevels.length == _flippedTopLevels.length);
+  }
+
+  void forEach(void process(TreeNode original, TreeNode flipped, int index)) {
+    for (int i = 0; i < _originalTopLevels.length; i++) {
+      process(_originalTopLevels[i], _flippedTopLevels[i], i);
+    }
+  }
+}
+
+
+class EditConfiguration {
+  final TextEditTransaction transaction;
```

Comment:
Because we use different transactions for depending on other configuration parameters.